File: 3/solution.jl
priorities = Dict(
    zip(vcat(collect('a':'z'), collect('A':'Z')), collect(1:52))
)

lines = readlines(ARGS[1])

# part 1
println(
    sum(
        map(
            ln -> sum(
                map(
                    x -> priorities[x],
                    collect(intersect(Set(ln[begin:length(ln) ÷ 2]), Set(ln[length(ln) ÷ 2 + 1:end]))),
                ),
            ),
            lines
        )
    )
)

# part 2
groups = [lines[i:i+2] for i in 1:3:length(lines)]

println(
    sum(
        map(
            y -> sum(
                map(
                    x -> priorities[x],
                    collect(intersect(Set(y[1]), Set(y[2]), Set(y[3])))
                )
            ),
            groups,
        )
    )
)

File: 3/solution.py
import sys
import string

priorities = dict([
    *zip(string.ascii_lowercase, range(1, 27)),
    *zip(string.ascii_uppercase, range(27, 53)),
])

lines = open(sys.argv[1], "r").read().strip().split("\n")

# part 1
print(
    sum(
        sum(
            priorities[x] for x in
            set(line[:int(len(line)/2)]) & set(line[int(len(line)/2):])
        )
        for line in lines
    )
)

# part 2
groups = (lines[i:i+3] for i in range(0, len(lines), 3))

print(
    sum(
        sum(
            priorities[x] for x in
            set(a) & set(b) & set(c)
        )
        for a, b, c in groups
    )
)
